enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,486 @@
+		if err != nil {
+			return xerrors.Errorf("determine current snapshot: %w", err)
+		}
+		if len(snapshot.Presets) == 0 {
+			logger.Debug(ctx, "no templates found with prebuilds configured")
+			return nil
+		}
+
+		// TODO: bounded concurrency? probably not but consider
+		var eg errgroup.Group
+		for _, preset := range snapshot.Presets {
+			ps, err := snapshot.FilterByPreset(preset.ID)
+			if err != nil {
+				logger.Warn(ctx, "failed to find preset snapshot", slog.Error(err), slog.F("preset_id", preset.ID.String()))
+				continue
+			}
+
+			if !preset.UsingActiveVersion && len(ps.Running) == 0 && len(ps.InProgress) == 0 {
+				logger.Debug(ctx, "skipping reconciliation for preset; inactive, no running prebuilds, and no in-progress operations",
+					slog.F("template_id", preset.TemplateID.String()), slog.F("template_name", preset.TemplateName),
+					slog.F("template_version_id", preset.TemplateVersionID.String()), slog.F("template_version_name", preset.TemplateVersionName),
+					slog.F("preset_id", preset.ID.String()), slog.F("preset_name", preset.Name))
+				continue
+			}
+
+			eg.Go(func() error {
+				// Pass outer context.
+				err = c.ReconcilePreset(ctx, *ps)
+				if err != nil {
+					logger.Error(ctx, "failed to reconcile prebuilds for preset", slog.Error(err), slog.F("preset_id", preset.ID))
+				}
+				// DO NOT return error otherwise the tx will end.
+				return nil
+			})
+		}
+
+		return eg.Wait()
+	})
+	if err != nil {
+		logger.Error(ctx, "failed to reconcile", slog.Error(err))
+	}
+
+	return err
+}
+
+func (c *StoreReconciler) WithReconciliationLock(ctx context.Context, logger slog.Logger, fn func(ctx context.Context, db database.Store) error) error {
+	// This tx holds a global lock, which prevents any other coderd replica from starting a reconciliation and
+	// possibly getting an inconsistent view of the state.
+	//
+	// The lock MUST be held until ALL modifications have been effected.
+	//
+	// It is run with RepeatableRead isolation, so it's effectively snapshotting the data at the start of the tx.
+	//
+	// This is a read-only tx, so returning an error (i.e. causing a rollback) has no impact.
+	return c.store.InTx(func(db database.Store) error {
+		start := c.clock.Now()
+
+		// TODO: use TryAcquireLock here and bail out early.
+		err := db.AcquireLock(ctx, database.LockIDReconcileTemplatePrebuilds)
+		if err != nil {
+			logger.Warn(ctx, "failed to acquire top-level reconciliation lock; likely running on another coderd replica", slog.Error(err))
+			return nil
+		}
+
+		logger.Debug(ctx, "acquired top-level reconciliation lock", slog.F("acquire_wait_secs", fmt.Sprintf("%.4f", c.clock.Since(start).Seconds())))
+
+		return fn(ctx, db)
+	}, &database.TxOptions{
+		Isolation:    sql.LevelRepeatableRead,
+		ReadOnly:     true,
+		TxIdentifier: "template_prebuilds",
+	})
+}
+
+// SnapshotState determines the current state of prebuilds & the presets which define them.
+// An application-level lock is used
```

Comment:
This comment might be incomplete? Otherwise it's missing a closing period.

Reply:
I updated comment. Application-level lock is not used anymore.